mapdrawgdal.c

```diff
@@ -333,24 +333,32 @@ int msDrawRasterLayerGDAL(mapObj *map, layerObj *layer, imageObj *image,
 
   /*
    * Set up the band selection.  We look for a BANDS directive in the
-   * the PROCESSING options.  If not found we default to red=1 or
-   * red=1,green=2,blue=3 or red=1,green=2,blue=3,alpha=4.
+   * the PROCESSING options.  If not found we default to grey=1, grey=1,alpha=2,
+   * red=1,green=2,blue=3 or red=1,green=2,blue=3,alpha>=4.
    */
 
   if( CSLFetchNameValue( layer->processing, "BANDS" ) == NULL ) {
+    const int gdal_band_count = GDALGetRasterCount( hDS );
     red_band = 1;
 
-    if( GDALGetRasterCount( hDS ) >= 4
-        && GDALGetRasterColorInterpretation(
-          GDALGetRasterBand( hDS, 4 ) ) == GCI_AlphaBand )
-      alpha_band = 4;
+    if( gdal_band_count >= 4 )
+    {
+        /* The alpha band is not necessarily the 4th one */
+        for( i = 4; i <= gdal_band_count; i ++ ) {
+            if( GDALGetRasterColorInterpretation(
+                GDALGetRasterBand( hDS, i ) ) == GCI_AlphaBand ) {
```

Comment:
> Are we sure that all GDAL drivers will correctly identify which band is the raster (edit: you mean alpha ?) band?

Not necessarily, but in most cases they should... There are different situations : 
* the alpha band is correctly identified as such, in which case this patch improves the situation by finding it even if not the 4th one. The past situation was that a R,G,B,unknown,Alpha raster was treated as RGB only. With this patch, the Alpha band will be taken into account
* there was an alpha band but the driver fails to identify it as such (R,G,B,Unknown). The patch doesn't change the situation. MapServer only considers RGB
* there was no alpha band but the driver wrongly identifies it as alpha. It it was a 4 band raster - R,G,B,A(wrong) - then the patch doesn't change the behaviour. If it was a 5 or more band, R,G,B,unkown,A(wrong) then the patch will pick up the 5th band as alpha, whereas previously it was ignored.

So I'd say the benefits of the patch outweights the drawbacks . The only "regression" would be a 5 band raster with a driver wrongly identifying a band >= 5 as alpha. rather uncommon. In that situation, people can either put an explicit BANDS directive as a workaround. Or sometimes create a .aux.xml file or wrap it in a VRT to fix the interpretation.

> Should we set band==4 as the default if no band was explicitely marked as being the alpha one 

I don't think that would be a good idea. You can have R,G,B,Unknown(for example Near Infra Red) products.

> (with the current patch, band 0 will be used in that case)

Not sure to follow you regarding that. "alpha" is initialized to 0, which means no alpha band, and in code later, this is properly dealt as the absence of an alpha band